mediacloud/mediawords/crawler/download/feed/ap.py

```diff
@@ -0,0 +1,207 @@
+    """
+    def __init__(self):
+
+        self.api_version = '1.1'
+        self.retry_limit = 5
+        config = get_config()
+
+        if 'associated_press' in config:
+            self.api_key = config['associated_press'].get('apikey')
+
+        if self.api_key is None:
+            logger.error("Could not load api key from configuration file.")
+            raise ValueError("API key configuration data missing for associated_press.")
+
+    def feed(self,**kwargs) -> dict:
+        """Feed API endpoint (Documentation: https://api.ap.org/media/v/docs/api/Search-and-Feed/#feed)
+
+        METHOD: GET
+
+        ENDPOINT PARAMETERS:
+
+        q: Query Expression
+
+        include,exclude: Parameters used to customize the fields returned in the response.
+
+        text_links: Specifies the format of the text renditions (stories, captions, scripts and shotlists) to return in the response.
+        For stories, the valid value is nitf (NITF) or anpa (ANPA 1312). For captions, scripts and shotlists, the valid value is nitf (NITF).
+        The value of all returns all available formats (this is the default).
+
+        page_size: The maximum number of items to return per page. The default is 10 items with a maximum of 100 per page.
+
+        versions: Specifies whether to return all available versions of the content item and all ANPA filings or only the latest
+        (the same story in the ANPA format may be filed multiple times; for example, with a different category code).
+
+        REQUEST HEADERS:
+
+        Accept-Encoding: Compresses the response to the gzip format. The valid value is gzip.
+
+        """
+        url = 'https://api.ap.org/media/v/content/feed'.format(self.api_version)
+        params = {}
+        params['apikey'] = self.api_key
+        params.update(kwargs)
+        feed_data = self._make_request(url,params)
+        return json.loads(feed_data)['data']
+
+    def content(self,path,**kwargs) -> dict:
+        """Content API endpoint (Documentation: https://api.ap.org/media/v/docs/api/Content-Item/)
+        Example: https://api.ap.org/media/v[{version}]/content/{item_id}?apikey={apikey}[{optional_parameters}]
+
+        METHOD: GET
+
+        ENDPOINT PARAMETERS:
+
+        qt: Unknown. They are present in the feed response but don't appear to be in the documentation
+
+        et: Unknown. Same as above.
+
+        REQUEST HEADERS:
+
+        Accept-Encoding: Compresses the response to the gzip format. The valid value is gzip.
+        """
+
+        url = 'https://api.ap.org/media/v/content/{}'.format(path)
+
+        params = {}
+        params['apikey'] = self.api_key
+        params.update(kwargs)
+        content_data = self._make_request(url,params)
+        return content_data
+
+    def _make_request(self,
+                      url: str,
+                      params: dict = None) -> str:
+        """Internal method for making API requests"""
+
+        retries = self.retry_limit
```

Comment:
we have a weird homebrew implementation of LWP in python so that our web fetching code in perl and python goes through the same code path.  we include support for retries in that code.  that code makes sure to set the user agent and from headers to the configured values and also has some media cloud custom hooks, such as logging every web request we make.

here's some example code:

```
import mediawords.web.user_agent

ua = mediawords.web.user_agent.UserAgent()
ua.set_timing([1, 2, 4, 8, 16, 32, 64, 128, 256])
response = ua.get(url)
if response.is_success:
    return response.decoded_content()
else:
    raise SomeError("Unable to fetch url '%s': %s" % (url, response.as_string()))
```